src/session.ts

```diff
@@ -1,29 +1,66 @@
 import { Context } from './context'
 import { Middleware } from './types'
 
-export function session<SessionData extends object>({
-  ttl = Infinity,
-  store = new Map<string, { expires: number; session: SessionData }>(),
-  getSessionKey = (ctx: Context) =>
-    ctx.from && ctx.chat && `${ctx.from.id}:${ctx.chat.id}`,
-} = {}): Middleware.ExtFn<Context, { session?: SessionData }> {
-  const ttlMs = ttl * 1000
+export interface Storage<T> {
+  makeKey: (ctx: Context) => Promise<string | undefined>
+  getItem: (name: string) => Promise<T | undefined>
+  setItem: (name: string, value: T) => Promise<void>
+  deleteItem: (name: string) => Promise<void>
+}
+
+export class MemorySessionStorage<T> implements Storage<T> {
+  private readonly ttl: number
+  private readonly store = new Map<string, { session: T; expires: number }>()
 
+  constructor(ttl = Infinity) {
+    this.ttl = ttl * 1000
+  }
+
+  async makeKey(ctx: Context): Promise<string | undefined> {
+    const fromId = ctx.callbackQuery?.from?.id ?? ctx.from?.id ?? null
+    const chatId = ctx.callbackQuery?.message?.chat.id ?? ctx.chat?.id ?? null
```

Comment:
```suggestion
    const fromId = ctx.from?.id
    const chatId = ctx.chat?.id
```

Reply:
why, not consider callback button?

Reply:
https://github.com/telegraf/telegraf/blob/ab49c1a6c419416bc57780d15d458364aba5d832/src/context.ts#L145 https://github.com/telegraf/telegraf/blob/ab49c1a6c419416bc57780d15d458364aba5d832/src/context.ts#L155

Reply:
OK